Naive and improved models writeup (#244)
ihmeuw/easylink: docs/source/user_guide/tutorials/getting_started.rst

Getting Started
===============

Introduction
============
EasyLink is a framework that allows users to build and run highly configurable record linkage pipelines. 
Its configurability enables users to "mix and match" different pieces of record 
linkage software by ensuring that each piece of the pipeline conforms to standard patterns. 

For example, users at the Census Bureau could easily evaluate whether using a more sophisticated "blocking" 
method would improve results in a certain pipeline, without having to rewrite the entire pipeline. For more 
information, read about the `motivation <https://easylink.readthedocs.io/en/latest/concepts/pipeline_schema/index.html#motivation>`_
behind EasyLink.

Overview
--------
This tutorial introduces EasyLink concepts and features by demonstrating the software's usage. Covered 
concepts include the EasyLink record linkage "pipeline schema", Easylink pipeline configuration, running 
pipelines, changing record linkage step implementations, changing input data, evaluating and comparing 
results, and more. 

> .. contents::

Reviewer: > Is there a way to have this TOC not include the page title?

Is that what the `:local:` option does?

https://docutils.sourceforge.io/docs/ref/rst/directives.html#contents-options

Audience
--------
This tutorial is intended for people familiar with record linkage practices, who are interested
in easily comparing linkage results across different methods. This tutorial will *not* cover information 
about linkage techniques or software and assumes the reader is familiar with these concepts and 

> ready to implement that knowledge using EasyLink.

Reviewer: +1 for making it less intimidating. I might go even further: While it's true that EasyLink is designed for people who already know about record linkage and are ready to apply that knowledge, I think (having just gone through the tutorial) you actually don't have to know anything about record linkage to follow along, since it implements everything for you and explains a few basic concepts along the way. It might inherently be a bit intimidating to newbies due to the complexity of EasyLink, but I don't think we should actively discourage people brand new to record linkage from trying it -- for someone who's motivated, this tutorial could be a great way to dive in and get their hands dirty.

Tutorial prerequisites
----------------------
`Install EasyLink <https://github.com/ihmeuw/easylink?tab=readme-ov-file#installation>`_ if you haven't already. 

The tutorial uses the `Splink <https://moj-analytical-services.github.io/splink/index.html>`_ Python package 
for record linkage implementations, which is included when you install EasyLink. Splink knowledge is not 
required to complete the tutorial but may be helpful when configuring Splink models.

Simulated input data
--------------------
Our first demonstration of running an EasyLink pipeline will configure a simple, "naive" record linkage

> model using Splink implementations. Our pipeline will link the

Reviewer: ```suggestion
model using Splink implementations. Our pipeline will link
```
Extra "the"

  For threshold 0.996, len(false_positives)=19; len(false_negatives)=158

We eliminated 30 false negatives compared to the naive results, thanks to our model linking more records with columns that 
are similar but don't exactly match.

Linking 2030 datasets using improved pipeline
=============================================
Finally, let's run this same "improved" pipeline, but using :download:`input_data_demo_2030.yaml` 
as the input YAML, which uses the SSA and W-2 datasets from 2030 rather than 
2020. Like we did for 2020, we'll create a ``2030`` directory and save :download:`input_file_ssa.parquet <2030/input_file_ssa.parquet>` and 
:download:`input_file_w2.parquet <2030/input_file_w2.parquet>` into it.

We can run the same pipeline on different data by changing only the input parameter::

  $ easylink run -p pipeline_demo_improved.yaml -i input_data_demo_2030.yaml -e environment_local.yaml -I /mnt/team/simulation_science/priv/engineering/er_ecosystem/images
  python print_fp_fn_w2_ssa.py results/2025_06_26_11_17_52 .996
  10345 true links
  For threshold 0.996, len(false_positives)=14; len(false_negatives)=149

We get similar, but not identical, results with the 2030 data.

> .. note::

Reviewer 2: Change to a "Conclusion" (or more fun name) heading

"In this tutorial, we've demonstrated..."

"Not everything EasyLink can do has been covered in this tutorial. EasyLink currently includes a few more implementations we haven't used here, can run pipelines on a computational cluster managed by Slurm [link] or distribute work using Apache Spark [link], and has additional flexibility in the pipeline schema that we haven't demonstrated."

  In its current state, EasyLink provides only one or two implementations for each step, does not yet have documentation 
  to support users in creating their own implementations, and is not yet stable enough to be recommended as a tool for production pipelines.
  However, interested users are encouraged to utilize the provided implementations to their full potential
  by creating more pipelines, changing how implementations are configured, and linking different datasets. 

  We hope to be able to add more features in the future, including:

  - Full suite of implementations reflecting a range of common record linkage techniques
  - Documentation supporting users in creating their own implementations
  - User-experience improvements, especially regarding writing pipeline specifications and implementations
  - Auto-parallel sections for processing large scale data

>   - Spark support

Reviewer 2: Can remove this

Success! Our pipeline has linked the input data and outputted the results, the clusters of records it found. We'll take a look 
at these results later and see how the model performed.

.. note:: 
   The pipeline output in its current state can be a little confusing. Note that the number assigned 
   to the slurm jobs is different than the order the jobs are executed in - these job IDs are 
   assigned by `Snakemake <https://snakemake.readthedocs.io/en/stable/>`_, a workflow manager
   that EasyLink uses to run your linkage pipeline. Also note that several input validation jobs will run before any actual 
   step implementations.

   Finally, despite the final output line containing the phrase "Missing output files", 
   this pipeline finished executing successfully. The "Reason" displayed in the output is explaining 
   why the job was run (the step inputs were ready but the output file did not yet exist), rather than 
   conveying an error message. We plan to improve these terminal outputs in the future.

Naive model - command line arguments
====================================

> This section will explain the command line argument and show the file we pass to each one, including the 

Reviewer 2: ```suggestion
This section will explain the command line arguments and show the file we pass to each one, including the 
```

if you ran the ``easylink run`` command on a cloud (e.g. EC2) instance that you were connected to with SSH, the implementations would run on that instance,
and so on.

Input data
----------
The ``--input-data`` (``-i``) argument to ``easylink run`` accepts a YAML file specifying a list 
of paths to files or directories containing input data to be used by the pipeline. 
We passed ``input_data_demo.yaml``, the contents of which are shown below::

  input_file_ssa: 2020/input_file_ssa.parquet
  input_file_w2: 2020/input_file_w2.parquet
  known_clusters: known_clusters.parquet

Here we have defined the locations of the three input files we will use: the 2020 versions of the 
``Social Security Administration`` and ``W2 & 1099`` datasets, and an empty ``known_clusters`` file, since no
clusters are known to us before running this pipeline. 

.. note::
    To meet the input specifications for :ref:`datasets` defined by the pipeline schema (see the next section),
    the ``SSA`` and ``W2`` datasets, after being generated by pseudopeople, were modified
    to add the required ``Record ID`` column. ``SSA`` death records were also removed, 

>     leaving only ``creation`` type records.

Reviewer: Were the death records also removed to meet the specifications for datasets, or was this just done to simplify things?

                      configuration:
                        LINK_ONLY: true
                        BLOCKING_RULES: "'l.first_name == r.first_name,l.last_name == r.last_name'"
                  evaluating_pairs:
                    implementation:
                      name: splink_evaluating_pairs
                      configuration:
                        LINK_ONLY: true
                        BLOCKING_RULES_FOR_TRAINING: "'l.first_name == r.first_name,l.last_name == r.last_name'"
                        COMPARISONS: "'ssn:exact,first_name:exact,middle_initial:exact,last_name:exact'"
                        PROBABILITY_TWO_RANDOM_RECORDS_MATCH: 0.0001  # == 1 / len(w2)
              links_to_clusters:
                implementation:
                  name: splink_links_to_clusters
                  configuration:
                    THRESHOLD_MATCH_PROBABILITY: 0.996
          updating_clusters:
            implementation:
              name: default_updating_clusters
      canonicalizing_and_downstream_analysis:
        implementation:
          name: dummy_canonicalizing_and_downstream_analysis

.. raw:: html

  </details>

> The pipeline specification follows the structure defined in the :ref:`pipeline_schema`, a very important

Reviewer: I don't think you can nest bold markup within a link, can you? Based on a quick test I did, I think this will print the asterisks instead of making the text bold.

> The pipeline specification follows the structure defined in the :ref:`pipeline_schema`, a very important

Reviewer 2: Darn! I guess we can't make this bold then 😞

part of EasyLink. The EasyLink pipeline **schema** enforces the standard patterns that linkage step implementations must 
follow, enabling easy configuration and swapping. 

It defines the steps of the record linkage pipeline, the inputs and outputs for each step, and the required formats for 
each input or output data file. 

> It also describes a set of operators which are used by the schema to allow customization 

Reviewer: Similarly, it's unclear what this "It" is referring to. If it's referring to the same thing as in the previous sentence, maybe this could just be combined into the above paragraph to make it more clear.

      - determining_exclusions:
          implementation:
            name: default_determining_exclusions
            configuration:
              INPUT_DATASET: input_file_w2
        removing_records:
          implementation:
            name: default_removing_records
            configuration:
              INPUT_DATASET: input_file_w2
  clustering:
    substeps:
      ...
  updating_clusters:
    implementation:
      name: default_updating_clusters

The last step shown, ``updating_clusters``, looks similar to ``canonicalization_and_downstream_analysis`` above; it simply chooses 
an implementation for the step using the ``name`` key. 

The substeps of ``clustering`` are hidden -- we'll look at them next. 

The complicated part is ``determining_exclusions_and_removing_records`` and its ``clones`` key:

The schema can define steps as :ref:`cloneable_sections`, which create 
multiple copies of that section and allow different implementations or inputs to be defined 
for each copy. We can see that the :ref:`entity_resolution_sub_steps` schema section defines
``determining_exclusions`` and ``removing_records`` as cloneable in the diagram 

> (blue dashed box).

Reviewer: I think it might be worth adding a sentence or two here about what the "determining exclusions" and "removing records" steps do. I think even for people familiar with record linkage, it might not be obvious why these are here. I eventually figured out that if I click on the "Entity resolution sub-steps" link above and then scroll down, I can find definitions of these steps. But on a cursory first pass, this is a prominent chunk of the yaml file whose purpose was unclear to me, which was a bit disorienting.

        PROBABILITY_TWO_RANDOM_RECORDS_MATCH: 0.0001  # == 1 / len(w2)

We see that ``pre-processing`` is another cloneable step, allowing us to select different pre-processing implementations for different
input datasets. In this case, we leave the ``w2`` dataset unchanged, while changing the ``middle_name`` column in the ``ssa`` dataset 
to a ``middle_initial`` column to match the ``w2`` data.

Finally, we will configure the two Splink implementations.

For ``splink_blocking_and_filtering``, we set::

    LINK_ONLY: true
    BLOCKING_RULES: "l.first_name == r.first_name,l.last_name == r.last_name"

The first variable instructs Splink to link records between datasets without de-depulicating within 
datasets.
The second is used by the Splink implementation to define which pairs of records 
will be considered as possible matches (only records with matching first or last names).

For ``splink_evaluating_pairs``, we set::

  LINK_ONLY: true
  BLOCKING_RULES_FOR_TRAINING: "l.first_name == r.first_name,l.last_name == r.last_name"
  COMPARISONS: "ssn:exact,first_name:exact,middle_initial:exact,last_name:exact"
  PROBABILITY_TWO_RANDOM_RECORDS_MATCH: 0.0001  # == 1 / len(w2)

> The first two variables are used similarly to the previous implementation. The third 

Reviewer 2: ```suggestion
The first two variables are used similarly to the previous implementation.
``BLOCKING_RULES_FOR_TRAINING`` is specifically used for `estimating parameters in the model <https://moj-analytical-services.github.io/splink/demos/tutorials/04_Estimating_model_parameters.html#estimating-with-expectation-maximisation>`_.
``COMPARISONS``
```

defines the columns which will be compared by the Splink model, and how Splink will evaluate
whether the column values match (exact comparisons). The fourth is a parameter used in training

> the model and making predictions (see the Splink docs for 
> `more info <https://moj-analytical-services.github.io/splink/api_docs/training.html#splink.internals.linker_components.training.LinkerTraining.estimate_parameters_using_expectation_maximisation>`__). 

Reviewer 2: ```suggestion
the model and making predictions
(`see the Splink docs for more info <https://moj-analytical-services.github.io/splink/demos/tutorials/04_Estimating_model_parameters.html#estimation-of-probability_two_random_records_match>`__). 
```

And that's the whole pipeline specification for our naive Splink model! Next let's take a look at the results from when we ran the 
pipeline earlier.

Naive model - results
=====================

Input and output data is stored in Parquet files. For example, to see our original records, 
we can view the contents of the input files listed in ``input_data_demo.yaml`` using Python:

.. code-block:: console

>   $ # Create/activate a conda environment if you don't want to install globally!
>   $ pip install pandas pyarrow

Reviewer 2: ```suggestion
  $ # Activate your EasyLink conda environment!
```

EasyLink requires pandas and pyarrow already!

  2            0_5          0_8          30  854-13-6295  ...                   00000       W2     2020         2
  3         0_5621       0_2289          46  674-27-1745  ...                   00000       W2     2020         3
  4         0_5623       0_2289          83  794-23-1522  ...                   00000       W2     2020         4
  ...          ...          ...         ...          ...  ...                     ...      ...      ...       ...
  9898     0_18936       0_7621          23  006-92-7857  ...                   00000       W2     2020      9898
  9899     0_18936       0_7621          90  006-92-7857  ...                   00000       W2     2020      9899
  9900     0_18937       0_7621           1  182-82-5017  ...                   00000     1099     2020      9900
  9901     0_18937       0_7621         105  182-82-5017  ...                   00000     1099     2020      9901
  9902     0_18939       0_7621           9  283-97-5940  ...                   00000       W2     2020      9902
  [9903 rows x 25 columns]

  >>> pd.read_parquet("known_clusters.parquet")
  Empty DataFrame
  Columns: [Input Record Dataset, Input Record ID, Cluster ID]
  Index: []

> It can also be useful to setup an alias to more easily preview parquet files. Add the following to your 
> ``.bash_aliases`` or ``.bashrc`` file, and restart your terminal.

Reviewer 2: ```suggestion
It can also be useful to set up an alias to more easily preview parquet files.
Run the following line to do so.
(If you want this alias to persist across terminal restarts, you can add it to your ``.bashrc`` or ``.bash_aliases`` in your home directory.)
```

.. code-block:: console

  $ pqprint results/2025_06_26_10_13_31/result.parquet 
        Input Record Dataset  Input Record ID               Cluster ID
  0           input_file_ssa             4610   input_file_ssa-__-4610
  1           input_file_ssa             4612   input_file_ssa-__-4612
  2           input_file_ssa             4613   input_file_ssa-__-4613
  3           input_file_ssa             4614   input_file_ssa-__-4614
  4           input_file_ssa             4615   input_file_ssa-__-4615
  ...                    ...              ...                      ...
  25178        input_file_w2             4496  input_file_ssa-__-11207
  25179       input_file_ssa            14652  input_file_ssa-__-14652
  25180       input_file_ssa             9980  input_file_ssa-__-14652
  25181        input_file_w2             5349  input_file_ssa-__-14652
  25182        input_file_w2             5350  input_file_ssa-__-14652

  [25183 rows x 3 columns]

As we can see, the pipeline has successfully outputted a ``Cluster ID`` for every 
input record it was able to link to another record for our probability threshold 

> of ``99.6%``. ``Cluster ID`` names are chosen by Splink based on the first record 

Reviewer 2: ```suggestion
of 99.6%. ``Cluster ID`` names are chosen by Splink based on the first record 
```

  ...                    ...             ...                  ...              ...           ...
  515790      input_file_ssa            8586        input_file_w2              943  3.526073e-04
  515791      input_file_ssa            8591        input_file_w2             3326  7.227902e-07
  515792      input_file_ssa            8595        input_file_w2             3369  7.227902e-07
  515793      input_file_ssa            8596        input_file_w2             6458  3.526073e-04
  515794      input_file_ssa            8597        input_file_w2             3248  7.227902e-07

  [515795 rows x 5 columns]

The record pairs displayed in the preview are all far below the match threshold, but the full results could 
be investigated further using ``pandas.read_parquet()`` in a Python session.

The Splink implementations in our pipeline also produce some diagnostic charts which can be useful 
for evaluating results, such as the :download:`match weights chart <naive_match_weights.html>` 
(`Splink docs <https://moj-analytical-services.github.io/splink/charts/match_weights_chart.html>`__) and 
:download:`comparison viewer tool <naive_comparison_viewer.html>` 

> (`Splink docs <https://moj-analytical-services.github.io/splink/charts/comparison_viewer_dashboard.html>`__). 

Reviewer 2: I really don't like how these links look, which I believe is related to @stevebachmeier's custom CSS here: https://github.com/ihmeuw/easylink/blob/d8c56c6e73bd7f2d5f0108e94aa8c3caa041d981/docs/source/_static/style.css#L4-L8

Somehow it is getting overridden, but only for the first word. I'd probably be in favor of just removing this custom CSS.

Author: Hmm, getting rid of those lines didn't seem to change anything. I deleted build/html and rebuilt local docs.

Download and run it::

  $ python print_fp_fn_w2_ssa.py results/2025_06_26_10_13_31 .996
  9292 true links
  For threshold 0.996, len(false_positives)=19; len(false_negatives)=188

In other words, with a threshold 
probability of 99.6%, out of 9,262 true links to be found, our model missed 19 (false negatives),
and additionally linked 188 pairs that shouldn't have been linked (false positives). 

Depending on our goals with the linked data, we might increase the threshold to reduce false positives,
at the cost of increased false negatives.
But this was a simple linkage model.
Let's improve it to see if we can get a better performance tradeoff!

Configuring an improved pipeline
================================
Next, let's modify our naive pipeline configuration YAML to try to improve our results. Primarily, we 
will change the ``COMPARISONS`` we pass to ``splink_evaluating_pairs`` to use flexible comparison 
methods rather than exact matches, allowing us to link records which have typos or other noise in them. We'll 
use a new pipeline configuration YAML, :download:`pipeline_demo_improved.yaml`, with these changes.

> In ``splink_evaluating_pairs``, we make the following changes:

Reviewer 2: ```suggestion
In ``splink_evaluating_pairs``, we make the following change:
```

  +  COMPARISONS: "ssn:levenshtein,first_name:name,middle_initial:exact,last_name:name"
     PROBABILITY_TWO_RANDOM_RECORDS_MATCH: 0.0001  # == 1 / len(w2)

``COMPARISONS`` now uses 
`Levenshtein <https://moj-analytical-services.github.io/splink/api_docs/comparison_library.html#splink.comparison_library.LevenshteinAtThresholds>`_
comparisons for ``ssn``, and 
`Name <https://moj-analytical-services.github.io/splink/api_docs/comparison_library.html#splink.comparison_library.NameComparison>`_
comparisons for ``first_name`` and ``last_name``, to link similar but not identical SSNs and names.

By re-running the pipeline with these changes and then running the evauation script, we can see how our results compare::

  $ easylink run -p pipeline_demo_improved.yaml -i input_data_demo.yaml -e environment_local.yaml -I /mnt/team/simulation_science/priv/engineering/er_ecosystem/images
  $ python print_fp_fn_w2_ssa.py results/2025_06_26_11_08_57 .996
  9292 true links
  For threshold 0.996, len(false_positives)=19; len(false_negatives)=158

We eliminated 30 false negatives compared to the naive results, thanks to our model linking more records with columns that 

> are similar but don't exactly match.

Reviewer: Maybe also say that the number of false positives, 19, stayed the same, so using improved comparisons found us more links without giving us more bad links, which is better than simply reducing the match threshold.

Linking 2030 datasets using improved pipeline
=============================================
Finally, let's run this same "improved" pipeline, but using :download:`input_data_demo_2030.yaml` 

> as the input YAML, which uses the ``ssa`` and ``w2`` datasets from 2030 rather than 

Reviewer 2: ```suggestion
as the input YAML, which uses the SSA and W-2 datasets from 2030 rather than 
```

> 2020. Like before, create a ``/2030`` directory and save :download:`input_file_ssa.parquet <2030/input_file_ssa.parquet>` and 
> :download:`input_file_w2.parquet <2030/input_file_w2.parquet>` to it -- these are the 2030 versions of the datasets.

Reviewer 2: ```suggestion
2020. Like we did for 2020, we'll create a ``2030`` directory and save :download:`input_file_ssa.parquet <2030/input_file_ssa.parquet>` and 
:download:`input_file_w2.parquet <2030/input_file_w2.parquet>` into it.
```

We can run the same pipeline on different data by changing only the input parameter::

  $ easylink run -p pipeline_demo_improved.yaml -i input_data_demo_2030.yaml -e environment_local.yaml -I /mnt/team/simulation_science/priv/engineering/er_ecosystem/images
  python print_fp_fn_w2_ssa.py results/2025_06_26_11_17_52 .996
  10345 true links
  For threshold 0.996, len(false_positives)=14; len(false_negatives)=149

> Our results are similar with the 2030 data!

Reviewer 2: ```suggestion
We get similar, but not identical, results with the 2030 data.
```

> Our results are similar with the 2030 data!

Reviewer 2: This ending feels abrupt. I think we need to have a closing paragraph about what we've demonstrated, what else EasyLink can do currently, and potential future directions for EasyLink.

Author: Added a closing paragraph, let me know any thoughts especially about what future directions/features we should mentoin

The complicated part is ``determining_exclusions_and_removing_records`` and its ``clones`` key:

> As described in the link above, the steps "determining exclusions and removing records" identify and remove

Reviewer 2: Which link are you referring to here? Might be better to just re-link

> records that can be excluded from this linking pass to save computational time, generally, because they have 

Reviewer 2: ```suggestion
records that can be excluded from this linking pass to save computational time, generally because they have 
```

    4       input_file_ssa_2020           16353   input_file_w2_2020             7434      0.00057
    ...                     ...             ...                  ...              ...          ...
    515790  input_file_ssa_2020           16309   input_file_w2_2020             7604      0.00057
    515791  input_file_ssa_2020           16310   input_file_w2_2020             7604      0.00057
    515792  input_file_ssa_2020           16311   input_file_w2_2020             7604      0.00057
    515793  input_file_ssa_2020           16312   input_file_w2_2020             7604      0.00057
    515794  input_file_ssa_2020           16313   input_file_w2_2020             7604      0.00057

    [515795 rows x 5 columns] 

The record pairs displayed in the preview are all far below the match threshold, but the results could 
be investigated further using ``pandas.read_parquet()`` in a Python session.

The Splink implementations in our pipeline also produce some diagnostic charts which can be useful 
for evaluating results, such as the :download:`match weights chart<naive_match_weights.html>` and 

> :download:`comparison viewer tool<naive_comparison_viewer.html>`. These charts are available in the 

Reviewer 2: Let's also add some links to the relevant Splink docs here

``diagnostics/splink_evaluating_pairs`` subdirectory of the results directory for each pipeline run.

Finally, since we generated the input datasets ourselves, and therefore know the ground truth of 
which records are truly links, lets see how our naive model performed. For a threshold match

> probability of ``.997`` (chosen using match rate evaluation metrics), out of ``9262`` true links, 

Reviewer 2: Add download link for mini script, and show how to run it which will output these numbers.

> probability of ``.997`` (chosen using match rate evaluation metrics), out of ``9262`` true links, 
> we can calculate that our model results contained ``260`` false positives and ``43`` false negatives, 
> with the rest of the links being accurate.

Reviewer 2: ```suggestion
probability of ``.997``, out of 9,262 true links to be found our model missed 43 (false negatives),
and additionally linked 260 pairs that shouldn't have been linked (false positives).
Depending on our goals with the linked data, we might increase the threshold to reduce false positives,
at the cost of increased false negatives.
But this was a simple linkage model.
Let's improve it to see if we can get a better performance tradeoff!
```

I don't think you need to specify how the threshold was chosen. My suggestion also makes the false negative/positive definitions a bit more explicit

Configuring an improved pipeline
================================

> Next, lets modify our naive pipeline configuration YAML to try to improve those results. Primarily, we 

Reviewer 2: ```suggestion
Next, let's modify our naive pipeline configuration YAML to try to improve our results. Primarily, we 
```

    PROBABILITY_TWO_RANDOM_RECORDS_MATCH: 0.0000625  # == 1 / len(ssa)
    THRESHOLD_MATCH_PROBABILITY: 0
    LINK_ONLY: true

``COMPARISONS`` now uses 
`Levenshtein <https://moj-analytical-services.github.io/splink/api_docs/comparison_library.html#splink.comparison_library.LevenshteinAtThresholds>`_
comparisons for ``ssn``, and 
`Name <https://moj-analytical-services.github.io/splink/api_docs/comparison_library.html#splink.comparison_library.NameComparison>`_
comparisons for ``first_name`` and ``last_name``, to link similar but not identical SSNs and names.

We also use a more accurate value for 
`PROBABILITY_TWO_RANDOM_RECORDS_MATCH <https://moj-analytical-services.github.io/splink/api_docs/training.html#splink.internals.linker_components.training.LinkerTraining.estimate_parameters_using_expectation_maximisation>`_.

By re-running the pipeline with these changes, we can see how our results compare::

    $ easylink run -p pipeline_demo_improved.yaml -i input_data_demo.yaml -e environment_local.yaml -I /mnt/team/simulation_science/priv/engineering/er_ecosystem/images
    
For a threshold match

> probability of ``.25`` (chosen using match rate evaluation metrics), out of ``9262`` true links, 

Reviewer 2: This probability will strike people as extremely low. Do we understand why this is happening?

Author: Not really :)

we can calculate that our model results contained ``236`` false positives and ``9`` false negatives, 
with the rest of the links being accurate.

The false negatives are significantly lower, thanks to our model linking more records with columns that 
are similar but don't exactly match. The false positives are only slightly lower, since ``216`` records, or
around 2%, are affected by 
`"borrow a social security number" <https://pseudopeople.readthedocs.io/en/latest/noise/column_noise.html#borrow-a-social-security-number>`_
pseudopeople noise. The ``SSA`` and ``W2`` dataset have extremely limited columns in common aside 

> from SSNs (first, middle initial and last), which makes it difficult to link these "borrowed SSN" records.

Reviewer 2: ... difficult _not_ to link ...

Changing inputs
===============

> Finally, lets run this same "improved" pipeline, but using :download:`input_data_demo_2030.yaml` 

Reviewer 2: ```suggestion
Finally, let's run this same "improved" pipeline, but using :download:`input_data_demo_2030.yaml` 
```

> as the input YAML, which uses the ``SSA`` and ``W2`` datasets from ``2030`` rather than 
> ``2020``::

Reviewer 2: ```suggestion
as the input YAML, which uses the ``SSA`` and ``W2`` datasets from 2030 rather than 
2020::
```

    $ easylink run -p pipeline_demo_improved_2030.yaml -i input_data_demo_2030.yaml -e environment_local.yaml -I /mnt/team/simulation_science/priv/engineering/er_ecosystem/images
    

> For a threshold match

Reviewer 2: Insert a paragraph before this one (once it's true!) saying the utility of this, i.e. that you can run the same pipeline on different data by changing only one parameter

  10345 true links
  For threshold 0.996, len(false_positives)=14; len(false_negatives)=149

We get similar, but not identical, results with the 2030 data.

.. note::

>   In its current state, EasyLink provides only one implementation for each step, does not yet have documentation 

Reviewer 2: ```suggestion
  In its current state, EasyLink provides only one or two implementations for each step, does not yet have documentation 
```

>   to support users in creating their own implementations, and is intended to be used for demonstration purposes 
>   only. However, interested users are encouraged to utilize the provided implementations to their full potential

Reviewer 2: ```suggestion
  to support users in creating their own implementations, and is not yet stable enough to be recommended as a tool for production pipelines.
  However, interested users are encouraged to utilize the provided implementations to their full potential
```
